src/content/plugins.md

```diff
@@ -0,0 +1,325 @@
+if it doesn't need to do anything because its inputs have no changed. We can use a couple of node packages to
+achieve this. [fs-tree-diff](https://www.npmjs.com/package/fs-tree-diff) is used to calculate a diff between the
+2 states of a directory, and [walk-sync](https://www.npmjs.com/package/walk-sync) which synchronously walks all
+the files in a directory and produces a list, which is passed to `fs-tree-diff`.
+
+Let's take a look:
+
+```js
+import Plugin from 'broccoli-plugin';
+import FSTree from 'fs-tree-diff';
+import walkSync from 'walk-sync';
+
+export class Cacher extends Plugin {
+  constructor (inputNodes, options) {
+    super(inputNodes, {
+      ...options,
+      persistentOutput: true,
+    });
+
+    this._previous = [];
+  }
+
+  _hasChanged() {
+    let changed = false;
+    for (let inputPath of this.inputPaths) {
+      const current = FSTree.fromEntries(walkSync.entries(inputPath));
+      const patch = current.calculatePatch(this._previous[inputPath] || []);
+      this._previous[inputPath] = current;
+
+      if (patch.length) {
+        changed = true;
+      }
+    }
+
+    return changed;
+  }
+
+  build() {
+    if (!this._hasChanged()) {
+      return;
+    }
+
+    // do work
+  }
+}
+
+export default function cacher(...params) {
+  return new Cacher(...params);
+}
+```
+
+Let's examine the above. First off we're importing the packages, nothing exciting there.
+In the `constructor()` we're setting `persistentOutput: true` in the options hash to true, this ensures that 
+Broccoli does not delete the output directory before calling `build()`, we need this so our output is preserved 
+and we can reuse it.
+
+```js
+  constructor (inputNodes, options) {
+    super(inputNodes, {
+      ...options,
+      persistentOutput: true,
+    });
+
+    this._previous = [];
+  }
+```
+
+Next, we've added a `_hasChanged()` method that does our cache state checking. 
+
+```js
+  _hasChanged() {
+    let changed = false;
+    for (let i in this.inputPaths) {
+      const current = FSTree.fromEntries(walkSync.entries(this.inputPaths[i]));
+      const patch = current.calculatePatch(this._previous[i] || []);
+      this._previous[i] = current;
+
+      if (patch.length) {
+        changed = true;
+      }
+    }
+
+    return changed;
+  }
+```
+
+Here, we're iterating each of `this.inputPaths` and using `FSTree.fromEntries()` and `walkSync.entries()` together 
+to produce a `current` state, then calculating a patch which tells us about added/removed/changed files using 
+`calculatePatch()`, and setting `changed` if there are any changes, then returning the `changed` result.
+
+```js
+  build() {
+    if (!this._hasChanged()) {
+      return;
+    }
+
+    // do work
+  }
+```
+
+Then in the `build()` method we simply call `this._hasChanged()` and if nothing has changed, we skip doing any
+further work. That's it, our plugin will now skip doing any work if none of the input files have changed.
```

Comment:
I loved this section on caching.

Maybe it's worth adding a mention as to whether you should be writing your own caching mechanism, or when it makes sense to use an approach provided by `broccoli-filter` and the other community caching plugins?

Reply:
Also, it's not clear whether the caching example is still expected to use the `walkSync` approach from earlier; I think if you just used the walkSync from before, the solution would not handle deleted files (i.e they'd be preserved in the output dir because of `persistentOutput:true`). If that's the case, we should mention that, but perhaps it'd also be better to not lose the diff information computed in `_hasChanged()`; if in fact the input tree has changed,  you'll want access to that diff again to figure out what needs to be deleted.

Reply:
Yeah i want to get into that in a separate update to this.
The general feeling amongst Broccoli maintainers is that we want to simplify the available base classes, and that caching is a cross-cutting concern that shouldn't need to be "inherited" from a base class.

IMO the base classes should consist of:

* `broccoli-plugin` => basic do what you like base, with the option to pass through files from the inputNode to output (if you wanted to just add logging, or write some new files to the output)
* `broccoli-filter` => 1:1 input to output file manipulations
* `broccoli-multi-filter` => M:1 inputs to output file manipulations (perhaps combine as an option with broccoli-filter)

> the solution would not handle deleted files 

The `_hasChanged()` should return true if input files are deleted, what you do with your output if up to you, if you need to delete files from outputPath then you must do that.

Reply:
But to know what files to delete, you have to do a diff again. And most beginners following this example won't remember this important corner case, and given how many Broccoli-doesn't-notice-my-deleted-files bugs we used to have, I think it'd be important to lead with a reasonable bullet proof example. As it stands, this reminds me of some of the documentation I've seen introducing beginners to async patterns in Ember action handlers using promises, setting people up for the inevitable corner case of a promise outliving the parent component and causing set-on-delete errors.

Reply:
Ah yeah, perhaps it should be `getChanges()` that returns the diff result? Then you could do whatever you want with it.

Reply:
Yeah, that's fine, and maybe comment below can mention the steps e.g. `// write the new/updated files, delete the removed files`